Jenkinsfile

```diff
@@ -144,16 +144,7 @@ Note this is another run, will double the time and no guaranty to have same erro
 
     post {
         always {
-            archiveArtifacts(artifacts: 'notebooks/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'pavics-sdi-*/docs/source/notebooks/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'pavics-sdi-*/docs/source/notebook-components/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'finch-*/docs/source/notebooks/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'raven-*/docs/source/notebooks/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'RavenPy-*/docs/notebooks/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'RavenPy-*/docs/notebooks/paper/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'esgf-compute-api-*/examples/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'PAVICS-landing-*/content/notebooks/climate_indicators/*.ipynb', fingerprint: true)
-            archiveArtifacts(artifacts: 'buildout/*.output.ipynb', fingerprint: true, allowEmptyArchive: true)
+            archiveArtifacts(artifacts: 'buildout/*.ipynb', fingerprint: true, allowEmptyArchive: true)
```

Comment:
Why not use a `buildout/**/*.ipynb` and leave the directory structure as originally defined by each repository? That would avoid all the directory renaming manipulations and potential side effects of removing pytest configs.

Reply:
I previously thought that too.  But actually real life usage, a flat directory is much easier to use, meaning to download the notebooks I want.  In a nested layout, if I want multiple notebooks under multiple different sub-folders, I have to click a lot more.  It's tiring.

Now with the option to only archive the notebooks enabled for the run, the chance to have name clash is even lower.

Reply:
> That would avoid all the directory renaming manipulations and potential side effects of removing pytest configs.

Not sure what you mean.  I do not perform any directory renaming manipulations.  The notebooks are run from the original location of the checkout because some will use relative path to get their test data in the same checkout.

I only copy the notebooks only, not the test data, to the `buildout/` folder so they are archived by Jenkins, along side their matching "output" notebook so we can download both of them and diff them manually if the diff presented by Jenkins console output is impossible to understand.

As for "potential side effects of removing pytest configs", I also do not understand how that relate to Jenkins archiving config change.

Reply:
All the notebooks are copied as a flat list, which could lead to conflicts between repositories. I was under the impression that the repository name was used as prefix to the archived notebook names to avoid this, but it is not the case. With the results archived this way, it is very hard to trace back the original source of the notebook.

![image](https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/assets/19194484/62d734b8-1218-49c3-a64d-306457c02115)

> As for "potential side effects of removing pytest configs", I also do not understand how that relate to Jenkins archiving config change.

It does not affect the archiving, but it affects how the notebooks are executed by `pytest` in order to generate their outputs. Therefore, if any `pytest` configurations options are applied to obtain certain effects needed by the tests, this could have bad side effects.

Reply:
> However, I'll add the repo name as prefix to the nb filename to help differentiate and avoid name clash.

01824907bcef64218b071d9a6504334c05780d60

Reply:
> Just so it is clear, it is already like this before (flat list of `*.output.ipynb`), I am just adding the original along-side the corresponding `output.ipynb`.

Older runs had the same nested directory structure as the source repositories, not a flat list. To me, that remains the clearest representation.

![image](https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/assets/19194484/7e17e641-3590-442d-8896-b154e9065a8a)

Reply:
> > Just so it is clear, it is already like this before (flat list of `*.output.ipynb`), I am just adding the original along-side the corresponding `output.ipynb`.
> 
> Older runs had the same nested directory structure as the source repositories, not a flat list. To me, that remains the clearest representation.
> 
> ![image](https://private-user-images.githubusercontent.com/19194484/331620356-7e17e641-3590-442d-8896-b154e9065a8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyNjkxODIsIm5iZiI6MTcxNjI2ODg4MiwicGF0aCI6Ii8xOTE5NDQ4NC8zMzE2MjAzNTYtN2UxN2U2NDEtMzU5MC00NDJkLTg4OTYtYjE1NGU5MDY1YThhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2Y2Y5ZDk5YjE5OWQ0YzU5YmI2YjhlNGRhZjllMDA1YWEzMDBhODBiMGYyNjJiMmIyZDhjMWQ4NGIyZTJjNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8Y8lo1TNBWHNftcURrZcI7_fwxzffqOgtRoVoojJHGo)

The old nested directory structure was for the original `.ipynb`, all the corresponding `.output.ipynb` are under `buildout/`.

Keeping the nested structure was a mistake I did initially when setting up this Jenkins.  Imagine you have weird diff outputs in 2 notebooks you can not understand so you have to download both the original `.ipynb` and the matching `.output.ipynb` files to manually diff them.  But these 2 notebooks are under different nested folders !  You'll have to click to get to the first older, back out, then again to the 2nd folder !  The more folders, the more back and forth it is !!!  

I've been in this annoying situation.  A flat list might seem odd initially but actually it's way more usable.  I have added the repo and branch as prefix to the filename.  I think it's a good compromise.

Reply:
I'd rather have this part of the code updated to save at the expected location under the relevant directory of the corresponding notebook.

https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/blob/d6fd5006337311095872418572384f20501f9dc6/runtest#L85-L97

Reply:
> I'd rather have this part of the code updated to save at the expected location under the relevant directory of the corresponding notebook.
> 
> https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/blob/d6fd5006337311095872418572384f20501f9dc6/runtest#L85-L97

Like already explained in comment https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/138#discussion_r1607653081, a flat list is faster and more productive than nested folder.  The risk of name collision has been addressed, per your request, by prefixing the filename with the repo and branch name.

You have tried this PR in comment https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/138#issuecomment-2115554022 and nothing was broken.

So, what else is bothering you?  Maybe we can work out some compromise to keep flat list?

Reply:
It's not a matter of breaking or not. I do not agree with the flat list. I find the original structure is more productive, because I usually have the repos cloned on my end (so I can push the relevant fix if needed), and mapping the paths between those and the outputs is more natural.

Reply:
> I usually have the repos cloned on my end (so I can push the relevant fix if needed), 

Same here.  If the notebook output change and I need to "refresh" the notebook, I download the `output.jpynb` version and override the matching original name.

This is the only workflow that involve committing and pushing I can see. 

This is precisely my workflow:

* right-click on the `output` file I want, select "Copy Link", I get something like http://jenkins.ouranos.ca/job/PAVICS-e2e-workflow-tests/job/current-production-version/302/artifact/buildout/finch-usage.output.ipynb

* then, `wget <full link I got> -O finch/docs/source/notebooks/finch-usage.ipynb`

* repeat for more files, right-click "copy link" gives me the full link so I do not have to manually edit the URL myself and risking to make edit mistake.

Can you show me exactly your workflow that involve committing and pushing?

> and mapping the paths between those and the outputs is more natural.

No matter what, there is no "direct" patch matching with your local checkout.

If we keep the nested list, the path would be something like http://.../artifacts/finch`-master`/docs/source/notebooks/finch-usage.`output.`ipynb.

So the branch name and the `output` in the path will always have to be removed.

Reply:
I don't bother copying/downloading any output because, usually, the errors need deeper investigation that require setup of another URL instance and rerunning the notebook. The CI instance is down already anyway, so no point targeting the output URL. So I simply run the notebook that is from the source repository, and compare with the corresponding path under the artifacts.

Reply:
> I simply run the notebook that is from the source repository, and compare with the corresponding path under the artifacts.

Why don't you just `git diff` and compare with your original checked out version?

Why compare with the original file saved under the artifacts and having to perform an additional `wget` to get that file?  That original file might not even be the same as your original checkout version!

Reply:
I've never really had any case where the output diff were not explicit enough in the log console to need another diff on my end. What I am more interested in are the complete output of the operations to understand the process flow as a whole, since those will tell be what went wrong. Simple issues like a single-line/value change are easily addressed with just the diff in the logs, so I don't need to diff them again.

> Why compare with the original file saved under the artifacts and having to perform an additional wget to get that file? That original file might not even be the same as your original checkout version!

Not sure what you mean. Like said, I do not do any `wget`. I just look at the output directly in the artifact and run the corresponding notebook on my end to investigate further the specific steps that caused problems when issues are more complicated than a simple diff.

Reply:
> This is exactly why I would rather have the notebooks with the generated outputs located under the same locations as the current artifacts using the same structure as displayed in the console outputs.

@fmigneault 
Okay, how about I make all the files (original and output) keep the same folder structure as prefix.  Then you have your structure to easy search for your file, I have my flat list to easily switch between multiples repos.

Something like `buildout/pavics-sdi-master--docs--source--notebooks--WCS_example{,-output]}.ipynb`.  Good enough compromise so we can merge this?

Reply:
I don't think it is a matter of compromise. The artifact structure should simply reflect the source repository. There is no need for an alternate naming convention and there is no need to define more code to handle it. The outputs should simply be saved where the notebooks are naturally collected in artifacts.

Reply:
> I don't think it is a matter of compromise.

Agreed, no compromise for usability.

Let me repeat comment https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/138#discussion_r1607653081 again if not clear enough: if "notebooks are under different nested folders ! You'll have to click to get to the first older, back out, then again to the 2nd folder ! The more folders, the more back and forth it is !!!"

Just a reminder, this artifact archiving just for immediate troubleshooting, it has no long term persistance value.  I'd rather focus on usability and productivity than following some standard for something that is ephemeral and not important.  Once you manage a lot of notebook repos, a flat list if much faster to deal with.

Reply:
Even then, I do not agree. If you know the path of the notebook you are interested in after getting an error, you have no reason to go back-and-forth between folders.

I see the addition of code, special character handling and deletion of repo-specific configs as a sign of bigger maintenance burden and potential side effects.